### ISSUE TITLE
Plain text slack webhooks or generic webhook support?

### ISSUE THREAD

**Issue by poVoq** [loomio/loomio#10111]
I am trying to send chat bot messages to an IRC chat room and have set up a bot that understands slack-like webhooks for it: https://github.com/jlu5/Limnoria-Grapnel

This seems to work generally fine, hoewever it looks like Loomio is not sending any plain text content at all and thus the bot is not triggered.

Maybe the option to "Send notification headline only" could be altered to use the plaintext functionality of the Slack like webhook?

There are also a few other system like XMPP that could probably benefit from this, so maybe a "generic" plain text only webhook option could be added in the chatbot menu?

Thanks.

**Comment by robguthrie** [comment]
Cool that you're trying this. I'm sure it will be easy to sort this out. 

Are you saying that you don't want to recieve a json message with a 'text' attribute (which we currently do), but rather just a text blob as the webhook payload?

**Comment by robguthrie** [comment]
Oh I just read the description of your repo. It is looking for a {text: 'blah'} formatted string. Which we do already. I think this is a configuration issue.

**Comment by poVoq** [comment]
Hmm, ok let me try again then, but I tried the same webhook with another slack-like webhook and it worked.

**Comment by poVoq** [comment]
So I can use the exact same webhook URL via curl as explained in the above linked repo and the "hello world" texts shows up fine on IRC, but from Loomio it doesn't work even though the "test connection" button seems to work fine and I get a "check your chatroom for a message from the bot” response.

I had a similar behaviour with slack-like webhooks from another Forum software and that seems to have been caused by a lack of plain text message in the json payload, but maybe here the problem is something else?

**Comment by poVoq** [comment]
Ok I think I narrowed it down.

There seem to be two issues:

First of all the json is send with nested double quotes, which I think is incorrect json. It should use outer single quotes instead. So instead of "{"text"}" rather '{"text”}'.

Also the send json is full of escaped quotes and other escaped characters which result in a failure to parse the json I think.

If I send the same json payload via curl but fix the above two issues manually it gets posted in the IRC chat as expected.

So maybe the officisl Slack json parser is able to fix those errors itself, but I think sending correct json would be better?

**Comment by robguthrie** [comment]
That's quite strange really. The json encoding is all handed by oj, a mature and well used gem. 

I have a suspicion that what you're describing is a result of the presentation code in your terminal rather than the actual json transmitted over the wire.

However I also doubt that suspicion because I'm sure you're looking deeply at the problem. So I dunno :) It's always possible I'm doing something dumb.

So my first question is what do I need to do to replicate this issue and confirm what you're seeing. I'll try to think out loud here and see if that helps.

I need to generate a message from loomio and see what is being sent.

I've setup a little test case that generates a webhook post, and I setup a nodejs server that echos the requests sent to it. (using code found here: https://gist.github.com/bszwej/62c327d773051816ed4949fd40c82c74)

Here's the request raw text I'm seeing:
```
==== POST /
> Headers
{
  'content-type': 'application/json; charset=utf-8',
  'accept-encoding': 'gzip;q=1.0,deflate;q=0.6,identity;q=0.3',
  accept: '*/*',
  'user-agent': 'Ruby',
  'sentry-trace': '72fa569049b340a8b3241ec1548f8aa6-3ded4c4e06334af2',
  baggage: 'sentry-trace_id=72fa569049b340a8b3241ec1548f8aa6,sentry-environment=test',
  host: 'localhost:8083',
  'content-length': '251'
}
> Body
{"text":"Herman Langosh started a thread\n\n**[Gov. Diedre Hand](http://localhost:3000/d/Q91TAaPH/gov-diedre-hand)**\n\nA description for this discussion. Should this be _rich_?\n\n\n\n\n","icon_url":"http://localhost","username":"Loomio","blocks":[]}
```

To me that seems about right. What do you think?

**Comment by robguthrie** [comment]
Just for extra info. I can also see the request being sent on the ruby side, but it's being encoded for screen so there is lots of extra escaped " characters. Which is what I think you're seeing on your tests. These are not actually being sent on the wire.

```
<- "POST / HTTP/1.1\r\nContent-Type: application/json; charset=utf-8\r\nAccept-Encoding: gzip;q=1.0,deflate;q=0.6,identity;q=0.3\r\nAccept: */*\r\nUser-Agent: Ruby\r\nSentry-Trace: 72fa569049b340a8b3241ec1548f8aa6-3ded4c4e06334af2\r\nBaggage: sentry-trace_id=72fa569049b340a8b3241ec1548f8aa6,sentry-environment=test\r\nHost: localhost:8083\r\nContent-Length: 251\r\n\r\n"
<- "{\"text\":\"Herman Langosh started a thread\\n\\n**[Gov. Diedre Hand](http://localhost:3000/d/Q91TAaPH/gov-diedre-hand)**\\n\\nA description for this discussion. Should this be _rich_?\\n\\n\\n\\n\\n\",\"icon_url\":\"http://localhost\",\"username\":\"Loomio\",\"blocks\":[]}"
```

**Comment by poVoq** [comment]
Ok you are probably right that it is a terminal rendering issue. Let me look further into it.

**Comment by robguthrie** [comment]
Cleaning up issues today. How are you going with this problem? 

I still think Loomio's json is valid, every other consumer of it has been working fine.

I had a look at the full error you posted, but it's not giving me any ideas, sorry!

**Comment by poVoq** [comment]
Currently no idea either. I guess it might be the Limnoria extension that is buggy.

**Comment by poVoq** [comment]
Looks like there is a fix in the bot extension now for "non-standard, but common behaviour", so hopefully it works now. Will try it  when I find the time today.

**Comment by poVoq** [comment]
This is the commit that makes the test messages work:
https://github.com/jlu5/Limnoria-Grapnel/commit/74eb1a9ac3c68ebe276837d321ac11802f1ece50

**Comment by robguthrie** [comment]
Oh.. maybe it's about the case of the header fields? They should downcase all headers and test that so case is ignored..
What error are you getting again?

**Comment by poVoq** [comment]
It seems to have been the missing multi-line support in Grapnel. Now that this was added it seems to be fully working.
